### PR TITLE
[Merged by Bors] - Rework ViewTarget to better support post processing

### PR DESCRIPTION
**Pull request by cart**
# Objective

Post processing effects cannot read and write to the same texture. Currently they must own their own intermediate texture and redundantly copy from that back to the main texture. This is very inefficient.

Additionally, working with ViewTarget is more complicated than it needs to be, especially when working with HDR textures. 

## Solution

`ViewTarget` now stores two copies of the "main texture". It uses an atomic value to track which is currently the "main texture" (this interior mutability is necessary to accommodate read-only RenderGraph execution). 

`ViewTarget` now has a `post_process_write` method, which will return a source and destination texture. Each call to this method will flip between the two copies of the "main texture".

```rust
let post_process = render_target.post_process_write();
let source_texture = post_process.source;
let destination_texture = post_process.destination;
```
The caller _must_ read from the source texture and write to the destination texture, as it is assumed that the destination texture will become the new "main texture".

For simplicity / understandability `ViewTarget` is now a flat type. "hdr-ness" is a property of the `TextureFormat`. The internals are fully private in the interest of providing simple / consistent apis. Developers can now easily access the main texture by calling `view_target.main_texture()`.

HDR ViewTargets no longer have an "ldr texture" with `TextureFormat::bevy_default`. They _only_ have their two "hdr" textures. This simplifies the mental model.  All we have is the "currently active hdr texture" and the "other hdr texture", which we flip between for post processing effects.

The tonemapping node has been rephrased to use this "post processing pattern". The blit pass has been removed, and it now only runs a pass when HDR is enabled. Notably, both the input and output texture are assumed to be HDR. This means that tonemapping behaves just like any other "post processing effect". It could theoretically be moved anywhere in the "effect chain" and continue to work.

In general, I think these changes will make the lives of people making post processing effects much easier. And they better position us to start building higher level / more structured  "post processing effect stacks".

---

## Changelog

- `ViewTarget` now stores two copies of the "main texture". Calling `ViewTarget::post_process_write` will flip between copies of the main texture.

### CODE REVIEW

File: crates/bevy_render/src/view/mod.rs [hunk 1]
```diff
@@ -116,78 +117,107 @@ pub struct ViewUniformOffset {
...
+    pub fn out_texture_format(&self) -> TextureFormat {
+        self.out_texture_format
+    }
+
+    /// This will start a new "post process write", which assumes that the caller
+    /// will write the [`PostProcessWrite`]'s `source` to the `destination`.
+    /// `source` is the "current" main texture. This will internally flip this
+    /// [`ViewTarget`]'s main texture to the `destination` texture, so the caller
+    /// _must_ ensure `source` is copied to `destination`, with or without modifications.
+    /// Failing to do so will cause the current main texture information to be lost.
```
Comment (dataphract) [inline crates/bevy_render/src/view/mod.rs:205]: Very minor nit, but having a [brief summary line](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html?highlight=summary#documenting-components) as a first paragraph makes this render more nicely in rustdoc.

**Review by dataphract — APPROVED** [review]
Looks great! Very excited to delete like a third of my JFA code :D

File: crates/bevy_render/src/view/mod.rs [hunk 2]
```diff
@@ -116,78 +117,107 @@ pub struct ViewUniformOffset {
...
-}
-
 #[derive(Component)]
 pub struct ViewTarget {
-    pub main_texture: ViewMainTexture,
-    pub out_texture: TextureView,
-    pub out_texture_format: TextureFormat,
+    main_textures: MainTargetTextures,
+    main_texture_format: TextureFormat,
+    /// 0 represents `texture_a`, 1 represents `texture_b`
```
Comment (devil-ira) [inline crates/bevy_render/src/view/mod.rs:124]:
```suggestion
    /// 0 represents `main_texture_a`, 1 represents `main_texture_b`
```

**Review by devil-ira — COMMENTED** [review]
I barely comprehend any of the rendering stuff but the diffs look nice :see_no_evil:
Just one nit :)

**Review by DGriffin91 — APPROVED** [review]

File: crates/bevy_render/src/view/mod.rs [hunk 3]
```diff
@@ -116,78 +117,107 @@ pub struct ViewUniformOffset {
...
+    main_texture: AtomicUsize,
```
Comment (robtfm) [inline crates/bevy_render/src/view/mod.rs:125]: should we force post processing to use &mut ViewTarget? i'm not sure but it feels like the atomic here could mask issues with concurrent writers, where there's no guarantee the render passes would be in the same order as the post_process_write requests.

**Review by IceSentry — APPROVED** [review]

File: crates/bevy_render/src/view/mod.rs (hunk 3 above)
Comment (cart) [reply crates/bevy_render/src/view/mod.rs:125]: Yeah this is currently valid. I would like to revisit what mutability might look like here though. There _is_ world where we choose to make node processing parallel. Thats one of the reasons we're restricting to `&World` in the first place (to make graph execution read-only).

I know a number of people (including myself) have explored the idea to move this stuff into systems. Still not sure how I feel about it, but it _would_ allow more granular read/writes in a parallel context (force linear command encoding for post processing effects, by nature of the mutable access to ViewTarget, allow parallel command encoding for most other things, as long as the systems are read only).

**Pull request closed without merging**